File: packages/vue-im-ui/src/utils/richDoc.ts
import { safeExternalUrl } from "../shared/contracts/url-safety";

/**
 * A rich-text message as it is drawn: the RichDoc v2 document the core stores (`docJson`, flare-proto
 * `RichTextContent.doc_json`) read into blocks and runs. The rule is shared with the three native kits
 * and written down in `spec/rich-doc-vectors.json`, which states it in this very shape.
 */
export type FlareRichMark = "bold" | "italic" | "underline" | "strike" | "spoiler";

export interface FlareRichRun {
  text: string;
  /** The known marks the run carries, once each, in the order of {@link FLARE_RICH_MARKS}. */
  marks?: FlareRichMark[];
  /** Inline code. */
  code?: true;
  /** The href of the innermost link around the run that `safeExternalUrl` accepts, as written. */
  link?: string;
  /** A mention: the user id it names, possibly empty. */
  mention?: string;
  /** An emoji: its pack key, possibly empty. */
  emoji?: string;
}

export type FlareRichBlock =
  | { kind: "paragraph"; runs: FlareRichRun[] }
  | { kind: "heading"; level: 1 | 2 | 3 | 4 | 5 | 6; runs: FlareRichRun[] }
  | { kind: "quote"; blocks: FlareRichBlock[] }
  | { kind: "code"; language?: string; text: string }
  | { kind: "list"; ordered: boolean; items: FlareRichBlock[][] }
  | { kind: "divider" };

export const FLARE_RICH_MARKS: readonly FlareRichMark[] = ["bold", "italic", "underline", "strike", "spoiler"];
/** The core's own limits (`rich_doc_v2/validate.rs`): no node deeper, no more nodes. */
export const FLARE_RICH_DOC_MAX_DEPTH = 64;
export const FLARE_RICH_DOC_MAX_NODES = 10_000;

type Node = Record<string, unknown>;

const asNode = (value: unknown): Node | null =>
  value !== null && typeof value === "object" && !Array.isArray(value) ? (value as Node) : null;
const childrenOf = (node: Node): unknown[] => (Array.isArray(node.children) ? node.children : []);
const nonEmpty = (value: unknown): string => (typeof value === "string" ? value : "");

/** Whether the tree stays inside the core's depth and node limits. Counted before anything is read. */
function withinLimits(root: Node): boolean {
  let nodes = 1;
  const stack: Array<[unknown, number]> = childrenOf(root).map((child) => [child, 1]);
  while (stack.length) {
    const [value, depth] = stack.pop()!;
    if (depth > FLARE_RICH_DOC_MAX_DEPTH) return false;
    nodes += 1;
    if (nodes > FLARE_RICH_DOC_MAX_NODES) return false;
    const node = asNode(value);
    if (node) for (const child of childrenOf(node)) stack.push([child, depth + 1]);
  }
  return true;
}

function readRuns(inlines: unknown[], link: string | undefined, out: FlareRichRun[] = []): FlareRichRun[] {
  const push = (run: FlareRichRun) => out.push(link === undefined ? run : { ...run, link });
  for (const value of inlines) {
    const node = asNode(value);
    if (!node) continue;
    switch (node.type) {
      case "text": {
        const text = nonEmpty(node.text);
        if (!text) break;
        const named = new Set(
          (Array.isArray(node.marks) ? node.marks : []).map((mark) => asNode(mark)?.type),
        );
        const marks = FLARE_RICH_MARKS.filter((mark) => named.has(mark));
        push(marks.length ? { text, marks } : { text });
        break;
      }
      case "inline_code": {
        const text = nonEmpty(node.text);
        if (text) push({ text, code: true });
        break;
      }
      case "hard_break":
        push({ text: "\n" });
        break;
      case "mention": {
        const userId = nonEmpty(node.user_id);
        const text = nonEmpty(node.text);
        if (userId || text) push({ text: text || `@${userId}`, mention: userId });
        break;
      }
      case "emoji": {
        const key = nonEmpty(node.key);
        const text = nonEmpty(node.text);
        if (key || text) push({ text: text || `:${key}:`, emoji: key });
        break;
      }
      case "link": {
        const href = nonEmpty(node.href);
        readRuns(childrenOf(node), href && safeExternalUrl(href) !== null ? href : link, out);
        break;
      }
      case "custom_inline":
        break;
      default: {
        const text = nonEmpty(node.text);
        if (text) push({ text });
      }
    }
  }
  return out;
}

function readBlocks(values: unknown[], out: FlareRichBlock[] = []): FlareRichBlock[] {
  for (const value of values) {
    const node = asNode(value);
    if (!node) continue;
    switch (node.type) {
      case "paragraph": {
        const runs = readRuns(childrenOf(node), undefined);
        if (runs.length) out.push({ kind: "paragraph", runs });
        break;
      }
      case "heading": {
        const runs = readRuns(childrenOf(node), undefined);
        if (!runs.length) break;
        const raw = typeof node.level === "number" && Number.isFinite(node.level) ? Math.trunc(node.level) : 1;
        out.push({ kind: "heading", level: Math.min(6, Math.max(1, raw)) as 1 | 2 | 3 | 4 | 5 | 6, runs });
        break;
      }
      case "quote": {
        const blocks = readBlocks(childrenOf(node));
        if (blocks.length) out.push({ kind: "quote", blocks });
        break;
      }
      case "code_block": {
        const text = childrenOf(node)
          .map((child) => {
            const inline = asNode(child);
            if (inline?.type === "text") return nonEmpty(inline.text);
            return inline?.type === "hard_break" ? "\n" : "";
          })
          .join("");
        if (!text) break;
        const language = nonEmpty(node.language);
        out.push(language ? { kind: "code", language, text } : { kind: "code", text });
        break;
      }
      case "bullet_list":
      case "ordered_list": {
        const items = childrenOf(node)
          .map(asNode)
          .filter((child): child is Node => child?.type === "list_item")
          .map((item) => readBlocks(childrenOf(item)))
          .filter((blocks) => blocks.length > 0);
        if (items.length) out.push({ kind: "list", ordered: node.type === "ordered_list", items });
        break;
      }
      case "divider":
        out.push({ kind: "divider" });
        break;
      case "custom_block":
        readBlocks(childrenOf(node), out);
        break;
      default:
        break;
    }
  }
  return out;
}

/**
 * What a covered spoiler run draws in place of `text`: every code point that is not whitespace becomes
 * U+3000 IDEOGRAPHIC SPACE, so the words are in neither the rendered text nor a copy until revealed.
 */
export function flareRichSpoilerCover(text: string): string {
  return Array.from(text, (character) => (/\s/.test(character) ? character : "\u3000")).join("");
}

/**
 * The blocks a rich-text body draws for `input` — the document object or its JSON text — or `null`
 * when it is not a drawable RichDoc v2 document, and the body shows the message's plain text instead.
 */
export function parseRichDoc(input: unknown): FlareRichBlock[] | null {
  let value = input;
  if (typeof input === "string") {
    try {
      value = JSON.parse(input);
    } catch {
      return null;
    }
  }
  const root = asNode(value);
  if (!root || root.type !== "doc" || root.version !== 2 || !Array.isArray(root.children)) return null;
  if (!withinLimits(root)) return null;
  return readBlocks(root.children);
}

File: packages/ios-im-ui/Sources/FlareIMUI/Models/RichDoc.swift
import Foundation

// A rich-text message as it is drawn: the RichDoc v2 document the core stores (`docJson`, flare-proto
// `RichTextContent.doc_json`) read into blocks and runs. The rule is shared with the other three kits and
// written down in `spec/rich-doc-vectors.json`; Vue implements it in `utils/richDoc.ts`.

/// The marks a run can carry, in the order a run lists them.
public enum FlareRichMark: String, CaseIterable, Sendable {
    case bold, italic, underline, strike, spoiler
}

/// One run of a paragraph or heading.
public struct FlareRichRun: Equatable, Sendable {
    public let text: String
    /// The known marks, once each, in ``FlareRichMark`` order.
    public let marks: [FlareRichMark]
    /// Inline code.
    public let code: Bool
    /// The href of the innermost link around the run that ``safeExternalURL(_:)`` accepts, as written.
    public let link: String?
    /// A mention: the user id it names, possibly empty.
    public let mention: String?
    /// An emoji: its pack key, possibly empty.
    public let emoji: String?

    public init(_ text: String, marks: [FlareRichMark] = [], code: Bool = false, link: String? = nil,
                mention: String? = nil, emoji: String? = nil) {
        self.text = text; self.marks = marks; self.code = code; self.link = link
        self.mention = mention; self.emoji = emoji
    }

    public func has(_ mark: FlareRichMark) -> Bool { marks.contains(mark) }

    fileprivate func linked(_ href: String?) -> FlareRichRun {
        guard let href else { return self }
        return FlareRichRun(text, marks: marks, code: code, link: href, mention: mention, emoji: emoji)
    }
}

/// A block of a rich-text body.
public indirect enum FlareRichBlock: Equatable, Sendable {
    case paragraph([FlareRichRun])
    /// `level` is 1…6.
    case heading(level: Int, runs: [FlareRichRun])
    case quote([FlareRichBlock])
    case code(text: String, language: String?)
    /// Each item's blocks; never empty.
    case list(ordered: Bool, items: [[FlareRichBlock]])
    case divider
}

/// What a covered spoiler run draws in place of `text`: every code point that is not whitespace becomes
/// U+3000 IDEOGRAPHIC SPACE, so the words are in neither the rendered text nor a copy until revealed.
public func flareRichSpoilerCover(_ text: String) -> String {
    var out = String.UnicodeScalarView()
    for scalar in text.unicodeScalars {
        out.append(scalar.properties.isWhitespace ? scalar : "\u{3000}")
    }
    return String(out)
}

/// The core's own limits (`rich_doc_v2/validate.rs`).
public let flareRichDocMaxDepth = 64
public let flareRichDocMaxNodes = 10_000

/// The blocks a rich-text body draws for `input` — the document's JSON text or its decoded dictionary —
/// or nil when it is not a drawable RichDoc v2 document, and the body shows the message's plain text.
public func flareParseRichDoc(_ input: Any?) -> [FlareRichBlock]? {
    var value = input
    if let text = input as? String {
        guard let data = text.data(using: .utf8),
              let decoded = try? JSONSerialization.jsonObject(with: data) else { return nil }
        value = decoded
    }
    guard let root = value as? [String: Any], root["type"] as? String == "doc",
          let version = RichDocReader.number(root["version"]), version == 2,
          let children = root["children"] as? [Any] else { return nil }
    guard RichDocReader.withinLimits(children) else { return nil }
    var out: [FlareRichBlock] = []
    RichDocReader.blocks(children, into: &out)
    return out
}

private enum RichDocReader {
    /// A JSON number, never a JSON boolean (which Foundation also hands over as an NSNumber).
    static func number(_ value: Any?) -> Double? {
        guard let number = value as? NSNumber, CFGetTypeID(number) != CFBooleanGetTypeID() else { return nil }
        return number.doubleValue
    }

    static func string(_ value: Any?) -> String { value as? String ?? "" }

    static func children(_ node: Any?) -> [Any] { (node as? [String: Any])?["children"] as? [Any] ?? [] }

    static func withinLimits(_ top: [Any]) -> Bool {
        var nodes = 1
        var stack: [(Any, Int)] = top.map { ($0, 1) }
        while let (node, depth) = stack.popLast() {
            if depth > flareRichDocMaxDepth { return false }
            nodes += 1
            if nodes > flareRichDocMaxNodes { return false }
            for child in children(node) { stack.append((child, depth + 1)) }
        }
        return true
    }

    static func runs(_ inlines: [Any], link: String?, into out: inout [FlareRichRun]) {
        func push(_ run: FlareRichRun) { out.append(run.linked(link)) }
        for value in inlines {
            guard let node = value as? [String: Any] else { continue }
            switch node["type"] as? String {
            case "text":
                let text = string(node["text"])
                guard !text.isEmpty else { break }
                let named = Set(((node["marks"] as? [Any]) ?? []).compactMap { ($0 as? [String: Any])?["type"] as? String })
                push(FlareRichRun(text, marks: FlareRichMark.allCases.filter { named.contains($0.rawValue) }))
            case "inline_code":
                let text = string(node["text"])
                if !text.isEmpty { push(FlareRichRun(text, code: true)) }
            case "hard_break":
                push(FlareRichRun("\n"))
            case "mention":
                let userId = string(node["user_id"]), text = string(node["text"])
                if !userId.isEmpty || !text.isEmpty { push(FlareRichRun(text.isEmpty ? "@\(userId)" : text, mention: userId)) }
            case "emoji":
                let key = string(node["key"]), text = string(node["text"])
                if !key.isEmpty || !text.isEmpty { push(FlareRichRun(text.isEmpty ? ":\(key):" : text, emoji: key)) }
            case "link":
                let href = string(node["href"])
                let accepted = !href.isEmpty && safeExternalURL(href) != nil
                runs(children(node), link: accepted ? href : link, into: &out)
            case "custom_inline":
                break
            default:
                let text = string(node["text"])
                if !text.isEmpty { push(FlareRichRun(text)) }
            }
        }
    }

    static func blocks(_ values: [Any], into out: inout [FlareRichBlock]) {
        for value in values {
            guard let node = value as? [String: Any] else { continue }
            switch node["type"] as? String {
            case "paragraph":
                var inline: [FlareRichRun] = []
                runs(children(node), link: nil, into: &inline)
                if !inline.isEmpty { out.append(.paragraph(inline)) }
            case "heading":
                var inline: [FlareRichRun] = []
                runs(children(node), link: nil, into: &inline)
                guard !inline.isEmpty else { break }
                let raw = number(node["level"]).flatMap { $0.isFinite ? Int($0.rounded(.towardZero)) : nil } ?? 1
                out.append(.heading(level: min(6, max(1, raw)), runs: inline))
            case "quote":
                var inner: [FlareRichBlock] = []
                blocks(children(node), into: &inner)
                if !inner.isEmpty { out.append(.quote(inner)) }
            case "code_block":
                let text = children(node).map { child -> String in
                    guard let inline = child as? [String: Any] else { return "" }
                    switch inline["type"] as? String {
                    case "text": return string(inline["text"])
                    case "hard_break": return "\n"
                    default: return ""
                    }
                }.joined()
                guard !text.isEmpty else { break }
                let language = string(node["language"])
                out.append(.code(text: text, language: language.isEmpty ? nil : language))
            case "bullet_list", "ordered_list":
                let items = children(node).compactMap { child -> [FlareRichBlock]? in
                    guard let item = child as? [String: Any], item["type"] as? String == "list_item" else { return nil }
                    var inner: [FlareRichBlock] = []
                    blocks(children(item), into: &inner)
                    return inner.isEmpty ? nil : inner
                }
                if !items.isEmpty { out.append(.list(ordered: node["type"] as? String == "ordered_list", items: items)) }
            case "divider":
                out.append(.divider)
            case "custom_block":
                blocks(children(node), into: &out)
            default:
                break
            }
        }
    }
}
